lib/history/sqlite/sqlite.go

```diff
@@ -30,115 +29,142 @@ import (
+	statusMutex sync.Mutex
 }
 
 // Config defines Timeline configuration.
 type Config struct {
 	// DBPath specifies the database location.
 	DBPath string
-	// Capacity specifies the max number of events that can be stored in the timeline.
-	Capacity int
+	// RetentionDuration specifies the duration to store events.
+	RetentionDuration time.Duration
 	// Clock will be used to record event timestamps.
 	Clock clockwork.Clock
 }
 
 // NewTimeline initializes and returns a new Timeline with the
 // specified configuration.
 func NewTimeline(ctx context.Context, config Config) (*Timeline, error) {
-	database, err := initSQLite(ctx, config.DBPath)
-	if err != nil {
-		return nil, trace.Wrap(err)
+	timeline := &Timeline{
+		config: config,
 	}
 
-	size, err := getSize(ctx, database)
-	if err != nil {
-		database.Close()
-		return nil, trace.Wrap(err)
+	if err := timeline.initSQLite(ctx); err != nil {
+		return nil, trace.Wrap(err, "failed to init sqlite")
 	}
 
-	return &Timeline{
-		Config:   config,
-		size:     size,
-		database: database,
-		// TODO: store and recover lastStatus in case satellite agent restarts.
-		lastStatus: nil,
-	}, nil
+	// if err := timeline.initPrevStatus(ctx); err != nil {
+	// 	return nil, trace.Wrap(err, "failed to init previous status")
+	// }
+
+	// TODO: should this be called in constructor method?
+	// Let caller run event eviction loop?
+	// Create separate init function?
+	go timeline.eventEvictionLoop()
+
+	return timeline, nil
 }
 
-// initSQLite initializes connection to database provided by dbPath and
-// initializes `events` table.
-func initSQLite(ctx context.Context, dbPath string) (*sqlx.DB, error) {
-	database, err := sqlx.ConnectContext(ctx, "sqlite3", dbPath)
+// initSQLite initializes connection to database and initializes `events` table.
+func (t *Timeline) initSQLite(ctx context.Context) error {
+	database, err := sqlx.ConnectContext(ctx, "sqlite3", t.config.DBPath)
 	if err != nil {
-		return nil, trace.Wrap(err)
+		return trace.Wrap(err)
 	}
 
 	if _, err := database.ExecContext(ctx, createTableEvents); err != nil {
-		return nil, trace.Wrap(err)
+		return trace.Wrap(err)
 	}
 
-	return database, nil
+	t.database = database
+	return nil
+}
+
+// initPrevStatus initializes the previously record status in case satellite
+// is restarted.
+func (t *Timeline) initPrevStatus(ctx context.Context) error {
+	// TODO
+	return trace.NotImplemented("not implemented")
 }
 
-// getSize returns the current number of rows stored in the provided database.
-func getSize(ctx context.Context, database *sqlx.DB) (size int, err error) {
-	row := database.QueryRowContext(ctx, "SELECT COUNT(*) FROM events")
-	if err := row.Scan(&size); err != nil {
-		return -1, trace.Wrap(err)
+// eventEvictionLoop periodically evicts old events to free up storage.
+func (t *Timeline) eventEvictionLoop() {
+	ticker := t.config.Clock.NewTicker(evictionFrequency)
+	for {
+		<-ticker.Chan()
```

Comment:
So I would rewrite this method a little bit and have it accept an external context and use select statement so the caller can potentially cancel it. Something like:

```go
func (t *Timeline) eventEvictionLoop(ctx context.Context) {
  ticker := ...
  defer ticker.Stop()
  for {
    select {
      <-ticker.C:
        localCtx, cancel := context.WithTimeout(ctx, evictionTimeout)
        // call t.evictEvents(localCtx, ...)
      <-ctx.Done():
        log.Info("Eviction loop is stopping.")
        return
    }
  }
}        
```

Reply:
The eviction loop is called from the timeline constructor method. In this situation should the constructor take an additional context parameter? Should the timeline struct initialize its own context to be used for the eviction loop and have some kind of shutdown method to cancel the context?